File: CanteenServiceDatabaseServer/GrpcService/converters/ReservationConverter.cs
using DTO;
using Google.Protobuf.Collections;
using grpcProtoFiles;

namespace GrpcService.converters;

public class ReservationConverter {
    private readonly UserConverter _userConverter;
    private readonly MenuConverter _menuConverter;

    public ReservationConverter(UserConverter userConverter, MenuConverter menuConverter) {
        _userConverter = userConverter;
        _menuConverter = menuConverter;
    }


    public ReservationDTO ToDto(Reservation reservation) {
        return new ReservationDTO() {
            Id = reservation.Id,
            ReservedBy = _userConverter.ToDto(reservation.ReservedBy),
            Menu = _menuConverter.ToDto(reservation.Menu),
            Quantity = reservation.Quantity,
            IsDelivered = reservation.IsDelivered
        };
    }

    public List<ReservationDTO> ToListDto(RepeatedField<Reservation> reservations) {
        return reservations.Select(ToDto).ToList();
    }

    public Reservation ToProto(ReservationDTO reservationDto) {
        return new Reservation() {
            Id = reservationDto.Id,
            ReservedBy = _userConverter.ToProto(reservationDto.ReservedBy),
            Menu = _menuConverter.ToProto(reservationDto.Menu),
            Quantity = reservationDto.Quantity,
            IsDelivered = reservationDto.IsDelivered
        };
    }

    public RepeatedField<Reservation> ToListProto(List<ReservationDTO> reservations) {
        return new RepeatedField<Reservation>(){reservations.Select(ToProto).ToList()};
    }
}

File: CanteenService/src/main/java/com/example/canteenservice/grpc/converters/ReservationConverter.java
package com.example.canteenservice.grpc.converters;

import com.example.canteenservice.dto.CreateReservationDTO;
import com.example.canteenservice.dto.ReservationDTO;
import grpcProtoFiles.ReservationOuterClass;
import org.springframework.stereotype.Component;

import java.util.List;

@Component
public class ReservationConverter {

    private final UserConverter userConverter;
    private final MenuConverter menuConverter;

    public ReservationConverter(UserConverter userConverter, MenuConverter menuConverter) {
        this.userConverter = userConverter;
        this.menuConverter = menuConverter;
    }

    public ReservationDTO toDTO(ReservationOuterClass.Reservation reservation){
        ReservationDTO reservationDTO = new ReservationDTO();
        reservationDTO.setId(reservation.getId());
        reservationDTO.setMenu(menuConverter.toDTO(reservation.getMenu()));
        reservationDTO.setReservedBy(userConverter.toDto(reservation.getReservedBy()));
        reservationDTO.setQuantity(reservation.getQuantity());
        reservationDTO.setDelivered(reservation.getIsDelivered());
        return reservationDTO;
    }

    public List<ReservationDTO> toDTOList(List<ReservationOuterClass.Reservation> reservations){
        // We map the list of reservations to a list of reservationDTOs
        return reservations.stream().map(this::toDTO).toList();
    }

    public ReservationOuterClass.Reservation toProto(ReservationDTO reservationDTO){
        return ReservationOuterClass.Reservation.newBuilder()
                .setId(reservationDTO.getId())
                .setMenu(menuConverter.toProto(reservationDTO.getMenu()))
                .setReservedBy(userConverter.toProto(reservationDTO.getReservedBy()))
                .setQuantity(reservationDTO.getQuantity())
                .setIsDelivered(reservationDTO.isDelivered())
                .build();
    }

}
